### PR TITLE
Joeri eisen :0

### PR DESCRIPTION
**Pull request by DonnyvanEijk**
Please watch

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,6 +8,19 @@ Also, it is important to know when you are using the actually API_URL  of accuwe
...
+We are using 2 API's on webserver to gather weather data from your location this API keys 
+will be needed to be placed by you bt yourself
+
+
+## Why a weather api?
+
+This is useful in all kinds of ways. From going to weather apps to websites with build in weather displayu
+There are enough things possible with this API
+
+
```
Comment (Joeri-Abbo) [inline README.md:22]: Missing the part about what routes are registered in the api

File: main.py [hunk 2]
```diff
@@ -1,129 +1,156 @@
...
+                weather_data['icon'],
+                weather_data['location']['lat'],
+                weather_data['location']['lng'],
+                weather_data['saved_at'],
+                weather_data['temperature']
+            ),
+            table='weather'
+        )
+    except Exception as e:
+        print(f"Something went wrong with inputting the value into the database \n Error: {e}")
```
Comment (Joeri-Abbo) [inline main.py]: If a type error is thrown or a insert error is not returning the error to the user. 9 out of 10 occasions you would not return the error to the user

File: main.py [hunk 3]
```diff
@@ -1,129 +1,156 @@
...
+        location_data = response.json()
+        coordinates = location_data.get('loc', '').split(',')
+
+        if not len(coordinates) == 2:
+            return None, None
+
+        latitude, longitude = map(float, coordinates)
+        return latitude, longitude
     except Exception as e:
+        print(f"Server not found! Error:{e}")
```
Comment (Joeri-Abbo) [inline main.py]: If a type error is thrown or a insert error is not returning the error to the user. 9 out of 10 occasions you would not return the error to the user

File: main.py [hunk 4]
```diff
@@ -1,129 +1,156 @@
...
 
     response = requests.get(api_url, params=params)
 
-    if response.status_code == 200:
-        location_data_unformatted = response.json()
-        location_data = [location_data_unformatted]
-        location_key = location_data[0]['Key']
-        api_url = f'{api_url_base}/currentconditions/v1/{location_key}'
+    if not response.status_code == 200:
+        print("There seems to be an issue")
```
Comment (Joeri-Abbo) [inline main.py]: TODO?

File: main.py [hunk 5]
```diff
@@ -1,129 +1,156 @@
...
+        return datetime.strptime(saved_at_str, '%Y-%m-%d %H:%M:%S')
+    except ValueError:
+        return None
 
 @app.route('/')
 def weather():
     create_weather_table()
     latitude, longitude = get_current_location()
 
+    # Check if data is in the cache
```
Comment (Joeri-Abbo) [inline main.py]: You need to fetch the stored data from the database instead of using a {} what if the service restarts.

File: main.py [hunk 6]
```diff
@@ -1,129 +1,156 @@
...
+    cached_data = weather_cache.get((latitude, longitude))
+
     if latitude is not None and longitude is not None:
-        weather_response = get_weather(latitude, longitude)
+        if cached_data and 'saved_at' in cached_data and isinstance(cached_data['saved_at'], str):
```
Comment (Joeri-Abbo) [inline main.py]: How can you return faster instead of using multiple if statements inside a ifstatement

File: main.py [hunk 7]
```diff
@@ -1,129 +1,156 @@
...
-            weather_data = weather_response.json()[0]
+        if cached_data and 'saved_at' in cached_data and isinstance(cached_data['saved_at'], datetime) and datetime.now() - cached_data['saved_at'] < timedelta(hours=1):
+            return jsonify(cached_data)
+        else:
+            weather_response = get_weather(latitude, longitude)
 
-            temperature = weather_data['Temperature']['Metric']['Value']
-            icon = weather_data['WeatherText']
+            if weather_response.status_code == 200:
+                weather_data = weather_response.json()[0]
```
Comment (Joeri-Abbo) [inline main.py]: what if json returns a empty response?

File: main.py [hunk 8]
```diff
@@ -1,129 +1,156 @@
...
+                        "lat": latitude,
+                        "lng": longitude
+                    },
+                    "saved_at": current_time
+                }
 
-            return jsonify(response_data)
-        else:
-            return jsonify({'error': 'Unable to retrieve weather data'})
+                # Update the cache
```
Comment (Joeri-Abbo) [inline main.py]: TODOs?

**Review by Joeri-Abbo — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (DonnyvanEijk) [reply README.md:22]: fixed

File: main.py (hunk 2 above)
Comment (DonnyvanEijk) [reply main.py]: fixed

File: main.py (hunk 3 above)
Comment (DonnyvanEijk) [reply main.py]: fixed

File: main.py (hunk 4 above)
Comment (DonnyvanEijk) [reply main.py]: fixed

File: main.py (hunk 6 above)
Comment (DonnyvanEijk) [reply main.py]: fixed

File: main.py (hunk 7 above)
Comment (DonnyvanEijk) [reply main.py]: done

File: main.py (hunk 8 above)
Comment (DonnyvanEijk) [reply main.py]: ?

File: main.py (hunk 5 above)
Comment (DonnyvanEijk) [reply main.py]: Done

File: README.md [hunk 9]
```diff
@@ -8,6 +8,25 @@ Also, it is important to know when you are using the actually API_URL  of accuwe
...
+## API routes used
+
+The api routes we have used are from the accuweather service. and the ipinfo service route to your host on port 5000 where you will be able to see this data
+also in the dotenv you may find 2 API_URL's u can comment the on out which you don't want to use. Also its important to know that the api url that is connected to accuweather actually grabs it from the API
+but the other url takes it  from a https server which doesnt take api requests.
+
```
Comment (Joeri-Abbo) [inline README.md:42]: What responses do you have in your api?

File: README.md (hunk 1 above)
Comment (Joeri-Abbo) [reply README.md:22]: This can be resolved!

File: main.py [hunk 10]
```diff
@@ -1,129 +1,171 @@
...
+
 def create_weather_table():
     create_table_query = """
     CREATE TABLE IF NOT EXISTS weather (
         id INT AUTO_INCREMENT PRIMARY KEY,
-        bikeable BOOLEAN,
         icon VARCHAR(255),
-        lat FLOAT,
-        lng FLOAT,
+        lat VARCHAR(255),
```
Comment (Joeri-Abbo) [inline main.py:26]: Why not use a float?

File: main.py [hunk 11]
```diff
@@ -1,129 +1,171 @@
...
+    try:
+        response = requests.get(api_url, params=params)
+
+        if not response.status_code == 200:
+            return jsonify({'error': 'Location search error'})
 
-    if response.status_code == 200:
-        location_data_unformatted = response.json()
-        location_data = [location_data_unformatted]
+        location_data = [response.json()]
```
Comment (Joeri-Abbo) [inline main.py:89]: What is json is not valid

File: main.py (hunk 5 above)
Comment (Joeri-Abbo) [reply main.py]: Remove the comments if is finished

File: main.py [hunk 12]
```diff
@@ -1,129 +1,171 @@
...
+    if latitude is None or longitude is None:
+        return jsonify({'error': 'Unable to retrieve current location coordinates'})
 
-            temperature = weather_data['Temperature']['Metric']['Value']
-            icon = weather_data['WeatherText']
+    # Check if data is in the cache
+    cached_data = weather_cache.get((latitude, longitude))
 
-            bikeable = temperature > 10
+    # Check if cached data is recent
```
Comment (Joeri-Abbo) [inline main.py]: Remove the comments if is finished

**Review by Joeri-Abbo — CHANGES_REQUESTED** [review]

File: main.py (hunk 5 above)
Comment (DonnyvanEijk) [reply main.py]: all done

File: main.py (hunk 10 above)
Comment (DonnyvanEijk) [reply main.py:26]: because for some kind of reason the query will break when using float for this value

File: README.md (hunk 9 above)
Comment (DonnyvanEijk) [reply README.md:42]: Fixed

File: main.py (hunk 10 above)
Comment (DonnyvanEijk) [reply main.py:26]: Breaks the query

File: main.py (hunk 11 above)
Comment (DonnyvanEijk) [reply main.py:89]: fixed

File: main.py (hunk 12 above)
Comment (DonnyvanEijk) [reply main.py]: done

**Pull request merged**